File: Script_BAO/BAO3_upxml_relation.py
#! usr/bin/python3
# vue que j'ai déjà adapter les étiquettes au moment où j'ai tranféré conll en xml, je ne peux pas utiliser la fonction findall

#commande dans le terminal : exemple
#python3 BAO3_upxml_relation.py ./BAO2/BAO2_Py_udpipe3210.xml "obj" 3210

import re
import sys
#from pathlib import Path


# notre fonction pour extraire la relation prend deux arguments: le fichier transféré en xml d'udpipe en tant que corpus, et le nom de la relation
#Buffer utilié::<data [^>]+>

def extract_relation(corpus, relation):
	#on crée un buffer pour stocker les pair de (key d'inedx,token/lemme) phrases
	phrase_buf = {}
	#un autre buffer en forme de liste de tuple (lemme/token_dep,id_gouv)
	relation_buf = []
	#on stock ici les coules de dep et gouv
	couples = set()
	#on ouvre le fichier d'entrée et on lit ligne par ligne, et on déoupe chaque phrase pour les mettre dans une liste des lignes :
	with open(corpus,"r",encoding = "UTF-8") as file:
		lignes = file.readlines()
	for ligne in lignes:
		if ligne.startswith("<element>"):
			fileds =re.findall("<data[^>]+>([^<]+)</data>",ligne)
			#j'ai en tout 10 collones dont 4 j'en n'ai pas besoin:
			idx, pos,lemma,token,id_gouv,relat = fileds
		 #hésitation : je prend le token comme ce que l'on a fait perl ou je prend le lemma? quel impact il y aura sur mon graphe? À tester:::::( lemma au final) 
			phrase_buf[idx]=lemma
			if relat == relation:
				relation_buf.append((lemma,id_gouv))
		if ligne=="</phrase>\n":#découpage de lecture
		#écriture des couples (id_gouv,lemme)
			for dep_lemma,id_gouv in relation_buf:
				couples.add((f"{phrase_buf[id_gouv]}",f"{dep_lemma}"))
		
		#remets les buffer en état initial pour le prochain tour
			relation_buf = []
			phrase_buf = {}
			
	fic.write("Le nombre de relation repérées:"+str(len(couples))+"\n\n")			
	for item in couples:
		fic.write(f"{item[0]} -{relation}-> {item[1]}\n")
		

	
					
  #  ---------------------------------------
if __name__=="__main__":
	corpus=sys.argv[1]
	#Le corpus est le premier argument de la liste d'arguments.
	relation=sys.argv[2]
	#La relation est le deuxième argument.
	#On donne le numéro de la rubrique dans la commande et il est le 3e argument.
	rub = sys.argv[3]
	#On ouvre notre fichier de sortie et dans son nom, on précise la rubrique, que le fichier de travail est celui étiqueté avec UDpipe (ud) et la relation. 
	fic=open(f"./BAO3/{rub} Relations/relation_{relation}_py.txt", "w", encoding="utf-8")
	#On lance la fonction extract_relation() sur notre corpus et la relation.
	print("Extraction de la relation :",relation)
	extract_relation(corpus, relation)
	fic.close()
	

File: Script_BAO/Bao3_upxml_relation.pl
#!/usr/bin/perl
#----------------------------------------------------------------------------------
<<DOC; 
En entrée : sortie UDPIPE formatée en XML + une relation syntaxique
# En sortie la liste triée des couples Gouverneur,Dependant en relation
321O
#perl Bao3_upxml_relation.pl ./BAO2/BAO2_Pl_udpipe3210.xml  "obj" 3210

DOC
#----------------------------------------------------------------------------------
use strict;
use utf8;
binmode (STDOUT,":utf8");
#-------------------------------------------------------------------------------------
my $rub =$ARGV[2];
my $relation = $ARGV[1];
my %dico_relations=();
# on découpe le texte par phrase (liste d'items annotés et potentiellement dépendants)
$/="</phrase>\n";
open my $upxml ,"<:encoding(utf8)","$ARGV[0]";
while (my $phrase=<$upxml>){	#-------------------------------------------------------------------------------------
# on traite chaque "paragraphe" en le decoupant "items"
	my @lignes = split(/\n/,$phrase);
	for (my $i=0; $i<=$#lignes; $i++) {
		# si la ligne lue contient la relation, on ira chercher le dep puis le gouv
		# on peut mettre les valeurs qu'on cherche sont entre () , et les numéroter en ordre
		if ($lignes[$i]=~/<element><data type="id">([^<]+)<\/data><data type="type">[^<]+<\/data><data type="lemma">[^<]+<\/data><data type="string">([^<]+)<\/data><data type="ID_GOV">([^<]+)<\/data><data type="relation">([^<]*$relation[^<]*)<\/data><\/element>/i){#attention aux positions des valeurs 
			my $id_dep=$1;
			my $token_dep=$2;
			my $id_gov= $3;
			my $relation=$4;
			#si le gouverneur se situe derrière le dépendant, on parcours les lignes suivantes et chercher la tête syntaxique
			if ($id_dep > $id_gov) {
				for (my $k=0; $k<$i; $k++) {
					if ($lignes[$k]=~/<element><data type="id">$id_gov<\/data><data type="type">[^<]+<\/data><data type="lemma">[^<]+<\/data><data type="string">([^<]+)<\/data><data type="ID_GOV">[^<]+<\/data><data type="relation">[^<]+<\/data><\/element>/) {
						my $token_gov=$1;
#On écrit les résultats dans le dico et on compte leur fréquence 
					$dico_relations{"$token_gov -$relation-> $token_dep"}++;
					}
				}
			}
			else {
				for (my $k=$i+1; $k<=$#lignes; $k++) {
					if ($lignes[$k]=~/<element><data type="id">$id_gov<\/data><data type="type">[^<]+<\/data><data type="lemma">[^<]+<\/data><data type="string">([^<]+)<\/data><data type="ID_GOV">[^<]+<\/data><data type="relation">[^<]+<\/data><\/element>/) {
						my $token_gov=$1;
						$dico_relations{"$token_gov -$relation-> $token_dep"}++;
					}
				}
			}
		}
	}
}
close ($upxml);

# on imprime la liste des couples Gouv,Dep et leur fréquence...
open my $extrait,">:encoding(utf8)", "./BAO3/$rub Relations/relation_obj_pl.txt";
my $cpt = 0;
foreach my $relation (sort {$dico_relations{$b}<=>$dico_relations{$a}} (keys %dico_relations)) {
	print $extrait "$dico_relations{$relation}\t$relation\n";
	$cpt++;
}
print $extrait "$cpt paires de relation ($relation) sont identifiées.";#À trouver à la fin du fichier :
close $extrait;
